# Проект спринта 7

Цели проекта
1. Изучение развития игровой индустрии: Проанализировать изменения и тенденции в игровой индустрии с 2000 по 2013 год.
2. Обзор платформ и жанров: Провести анализ объемов продаж игр различных жанров и региональных предпочтений игроков.

Задачи проекта
1. Сбор и проверка данных:
    - Ознакомиться с историческими данными о продажах игр, жанрах и платформах.
    - Проверить данные на корректность и полноту.
2. Фильтрация данных:
    - Отобрать игры, выпущенные в период с 2000 по 2013 год включительно.
3. Категоризация оценок:
    Разделить игры на три категории по оценкам пользователей и экспертов:
        - Высокая оценка: от 8 до 10 и от 80 до 100 (включительно).
        - Средняя оценка: от 3 до 8 и от 30 до 80 (не включая правые границы).
        - Низкая оценка: от 0 до 3 и от 0 до 30 (не включая правые границы).
4. Анализ платформ:
    - Выделить топ-7 платформ по количеству выпущенных игр за период с 2000 по 2013 год.

Описание данных

- Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
    - Name — название игры.
    - Platform — название платформы.
    - Year of Release — год выпуска игры.
    - Genre — жанр игры.
    - NA sales — продажи в Северной Америке (в миллионах проданных копий).
    - EU sales — продажи в Европе (в миллионах проданных копий).
    - JP sales — продажи в Японии (в миллионах проданных копий).
    - Other sales — продажи в других странах (в миллионах проданных копий).
    - Critic Score — оценка критиков (от 0 до 100).
    - User Score — оценка пользователей (от 0 до 10).
    - Rating — рейтинг организации ESRB (англ. Entertainment Software Rating Board). 

# Оглавление
- Загрузка и знакомство с данными
- Проверка ошибок в данных и их предобработка
- Фильтрация данных
- Категоризация данных

## Загрузка и знакомство с данными

### 1. Загрузите необходимые библиотеки Python и данные датасета

In [1]:
# импортируем библиотеки
import pandas as pd

In [2]:
try:
    df = pd.read_csv('/Users/vinnie/local_storage/Практикум/new_games.csv')
except FileNotFoundError:
    df = pd.read_csv('/datasets/new_games.csv')
df = pd.DataFrame(df)
df

Name Platform  Year of Release         Genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       NA sales EU sales JP sales  Other sales  Critic Score User Score Rating  
0         41.36    28.96     3.77         8.45          76.0          8      E  
1         29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2         15.68    12.76     3.79         3.29          82.0        8.3      E  
3         15.61    10.93     3.28         2.95          80.0          8      E  
4         11.27     8.89    10.22         1.00           NaN        NaN    NaN  
...         ...      ...      ...          ...           ...        ...    ...  
16951      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16952      0.00     0.01      0.0         0.00           NaN        NaN    NaN  
16953      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16954      0.01      0.0      0.0         0.00           NaN        NaN    NaN  
16955      0.00      0.0     0.01         0.00           NaN        NaN    NaN  

[16956 rows x 11 columns]

### 2. Познакомьтесь с данными: выведите первые строки и результат метода info().

In [3]:
# Выводим первые строки (5 шт)
df.head(5)

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

In [4]:
# результат метода info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [5]:
# подсчет исходного количества строк
rows_cnt = len(df)
rows_cnt

16956

### 3. Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.

1. Датафрейм содержит 16956 строк и 11 столбцов
2. Данные соответсвуют описанию
3. Полные данные представлены только в столбцах: Platform, NA sales, EU sales, JP sales, Other sales, остальные столбцы содержат пропуски. Меньше всего данных в столбце Critic Score (49% от общего количества), далее User Score (60% от общего количества), далее Rating (59% от общего количества). Для столбцов Name, Year of Release, Genre пропуски составляют менее 2%
4. Тип данных неверный для столбца "Year of Release" - тип данных должен быть без плавающей точки. Для столбцов EU sales, JP sales, User Score тип данных должен быть численный.

### 4. Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

Колонку Rating следует переименовать с указанием организации, которая дала таковую оценкую. Например, Rating ESRB

## Проверка ошибок в данных и их предобработка

### 1. Названия, или метки, столбцов датафрейма:

In [6]:
# Выведите на экран названия всех столбцов датафрейма
columns = df.columns
columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [7]:
# Проверка стиля написания
for column in columns:
    if column.islower():
        print(f"{column}: все буквы строчные")
    elif column.isupper():
        print(f"{column}: все буквы заглавные")
    elif column.istitle():
        print(f"{column}: слово начинается с заглавной буквы")
    else:
        print(f"{column}: смешанный стиль")

Name: слово начинается с заглавной буквы
Platform: слово начинается с заглавной буквы
Year of Release: смешанный стиль
Genre: слово начинается с заглавной буквы
NA sales: смешанный стиль
EU sales: смешанный стиль
JP sales: смешанный стиль
Other sales: смешанный стиль
Critic Score: слово начинается с заглавной буквы
User Score: слово начинается с заглавной буквы
Rating: слово начинается с заглавной буквы


In [8]:
# Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания _.
df.columns = df.columns.str.replace(' ', '_').str.lower()
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

### 2. Типы данных:

Если встречаются некорректные типы данных, предположите их причины.
1. Year of Release - год записывается с помощью чисел, поэтому могли ошибочно присвоить ему такой тип
2. В столбцах EU sales, JP sales и User Score тип данных "object" мог появится в результате внесения чисел с десятичной частью с разными знаками разделителей: "." или ","

In [9]:
# Удаляем все пропуски в столбце year_of_release (допустимо, их менее 2%)
df = df.dropna(subset=['year_of_release'])

In [10]:
# Изменяем тип данных в столбце year_of_release
df.loc[:, 'year_of_release'] = df['year_of_release'].astype('int64')
df

name platform  year_of_release         genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       na_sales eu_sales jp_sales  other_sales  critic_score user_score rating  
0         41.36    28.96     3.77         8.45          76.0          8      E  
1         29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2         15.68    12.76     3.79         3.29          82.0        8.3      E  
3         15.61    10.93     3.28         2.95          80.0          8      E  
4         11.27     8.89    10.22         1.00           NaN        NaN    NaN  
...         ...      ...      ...          ...           ...        ...    ...  
16951      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16952      0.00     0.01      0.0         0.00           NaN        NaN    NaN  
16953      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16954      0.01      0.0      0.0         0.00           NaN        NaN    NaN  
16955      0.00      0.0     0.01         0.00           NaN        NaN    NaN  

[16681 rows x 11 columns]

In [11]:
# преобразуем данные остальных столбцов к строковому типу, меняем символы и прeобразуем к численному типу
df.loc[:, 'eu_sales'] = pd.to_numeric(df['eu_sales'].astype('str').str.replace(',', '.'), errors='coerce')
df.loc[:, 'jp_sales'] = pd.to_numeric(df['jp_sales'].astype('str').str.replace(',', '.'), errors = 'coerce')
df.loc[:, 'user_score'] = pd.to_numeric(df['user_score'].astype('str').str.replace(',', '.'), errors = 'coerce')
df

name platform  year_of_release         genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       na_sales eu_sales jp_sales  other_sales  critic_score user_score rating  
0         41.36    28.96     3.77         8.45          76.0        8.0      E  
1         29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2         15.68    12.76     3.79         3.29          82.0        8.3      E  
3         15.61    10.93     3.28         2.95          80.0        8.0      E  
4         11.27     8.89    10.22         1.00           NaN        NaN    NaN  
...         ...      ...      ...          ...           ...        ...    ...  
16951      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16952      0.00     0.01      0.0         0.00           NaN        NaN    NaN  
16953      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16954      0.01      0.0      0.0         0.00           NaN        NaN    NaN  
16955      0.00      0.0     0.01         0.00           NaN        NaN    NaN  

[16681 rows x 11 columns]

In [12]:
# Проверка 
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16681 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16681 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16679 non-null  object 
 4   na_sales         16681 non-null  float64
 5   eu_sales         16675 non-null  object 
 6   jp_sales         16677 non-null  object 
 7   other_sales      16681 non-null  float64
 8   critic_score     8085 non-null   float64
 9   user_score       7558 non-null   object 
 10  rating           9901 non-null   object 
dtypes: float64(4), object(7)
memory usage: 1.5+ MB


### 3. Наличие пропусков в данных:

In [13]:
# Выводим абсолютное количество пустых значений в каждой колонке
df.isna().sum() 

name                  2
platform              0
year_of_release       0
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8596
user_score         9123
rating             6780
dtype: int64

In [14]:
# Выводим относительное количество пустых значение в каждой колонке с 2 знаками после запятой
round(df.isna().sum() / len(df) * 100, 2)

name                0.01
platform            0.00
year_of_release     0.00
genre               0.01
na_sales            0.00
eu_sales            0.04
jp_sales            0.02
other_sales         0.00
critic_score       51.53
user_score         54.69
rating             40.65
dtype: float64

Вывод
1. В колонках eu_sales, jp_sales можно либо ничего не делать и оставить все как есть, так как количество пропусков менее процента, либо заменить на среднее значение в зависимости от названия платформы и года выхода игры.
2. В столбцах critic_score, user_score заполнить пропуски агригированным значением было бы не правильно, так как у нас все фильмы разные и оценки у них могут быть разные или вообщее не быть. Удалить пропуски мы не можем, так как их более 40%. Для корректной работы кода можно заменить пропуск на численное значение-индикатор, которое сразу будет нам говорить о том что данных нет, например, на -1 - оценка не может быть отрицательной.
3. В колонках name, genre, rating заменяем на строковое значение "NS"

Причины возникновения пропусков в столбцах: 
- name: невнимательность сотрудников. Не может быть такого, чтобы у игры не было названия. Это подтверждает и доля пропусков.
- genre: невнимательность сотрудников. Не может быть такого, чтобы у игры не было жанра. Это подтверждает и доля пропусков.
- eu_sales и jp_sales: отсутствие продаж соответсствующей игры в данных странах.
- critic_score, user_score, rating - критики/пользователи/ организация просто не дали оценку игре

In [15]:
# Если df сам является срезом, работаем с его копией
df = df.copy()

df['user_score'] = pd.to_numeric(df['user_score'], errors='coerce').fillna(-1)
df['critic_score'] = pd.to_numeric(df['critic_score'], errors='coerce').fillna(-1)
df

name platform  year_of_release         genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       na_sales eu_sales jp_sales  other_sales  critic_score  user_score  \
0         41.36    28.96     3.77         8.45          76.0         8.0   
1         29.08     3.58     6.81         0.77          -1.0        -1.0   
2         15.68    12.76     3.79         3.29          82.0         8.3   
3         15.61    10.93     3.28         2.95          80.0         8.0   
4         11.27     8.89    10.22         1.00          -1.0        -1.0   
...         ...      ...      ...          ...           ...         ...   
16951      0.00      0.0     0.01         0.00          -1.0        -1.0   
16952      0.00     0.01      0.0         0.00          -1.0        -1.0   
16953      0.00      0.0     0.01         0.00          -1.0        -1.0   
16954      0.01      0.0      0.0         0.00          -1.0        -1.0   
16955      0.00      0.0     0.01         0.00          -1.0        -1.0   

      rating  
0          E  
1        NaN  
2          E  
3          E  
4        NaN  
...      ...  
16951    NaN  
16952    NaN  
16953    NaN  
16954    NaN  
16955    NaN  

[16681 rows x 11 columns]

In [16]:
# Обработайтка пропущенных значений пункт 1
def fill_columns_mean (row):
    group = df[(df['platform'] == row['platform']) & (df['year_of_release'] == row['year_of_release'])]
    return group[['eu_sales', 'jp_sales']].mean()
df.fillna(df.apply(fill_columns_mean, axis = 1))

name platform  year_of_release         genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       na_sales eu_sales jp_sales  other_sales  critic_score  user_score  \
0         41.36    28.96     3.77         8.45          76.0         8.0   
1         29.08     3.58     6.81         0.77          -1.0        -1.0   
2         15.68    12.76     3.79         3.29          82.0         8.3   
3         15.61    10.93     3.28         2.95          80.0         8.0   
4         11.27     8.89    10.22         1.00          -1.0        -1.0   
...         ...      ...      ...          ...           ...         ...   
16951      0.00      0.0     0.01         0.00          -1.0        -1.0   
16952      0.00     0.01      0.0         0.00          -1.0        -1.0   
16953      0.00      0.0     0.01         0.00          -1.0        -1.0   
16954      0.01      0.0      0.0         0.00          -1.0        -1.0   
16955      0.00      0.0     0.01         0.00          -1.0        -1.0   

      rating  
0          E  
1        NaN  
2          E  
3          E  
4        NaN  
...      ...  
16951    NaN  
16952    NaN  
16953    NaN  
16954    NaN  
16955    NaN  

[16681 rows x 11 columns]

In [17]:
# Обработайтка пропущенных значений пункт 3
df.loc[:, 'name'] = df['name'].fillna('NS')
df.loc[:, 'genre'] = df['genre'].fillna('NS')
df.loc[:, 'rating'] = df['rating'].fillna('NS')
df.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  eu_sales jp_sales  other_sales  critic_score  user_score rating  
0    28.96     3.77         8.45          76.0         8.0      E  
1     3.58     6.81         0.77          -1.0        -1.0     NS  
2    12.76     3.79         3.29          82.0         8.3      E  
3    10.93     3.28         2.95          80.0         8.0      E  
4     8.89    10.22         1.00          -1.0        -1.0     NS

### 4. Явные и неявные дубликаты в данных:

Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.

In [18]:
# Выводим количество уникальных строк в столбце name 
df['name'].nunique()

11427

In [19]:
# Выводим количество уникальных строк в столбце genre 
df['genre'].nunique()

25

In [20]:
# Выводим количество уникальных строк в столбце rating 
df['rating'].nunique()

9

In [21]:
# Преобразуем значения в столбцe name
df['name'].str.lower().nunique()

11427

In [22]:
# Преобразуем значения в столбцe genre
df['genre'].str.lower().nunique()

13

In [23]:
# Преобразуем значения в столбцe genre
df['rating'].str.upper().nunique()

9

In [24]:
# Выявление явных полных дубликатов в столбцах
df.duplicated(subset=None, keep='first') 

0        False
1        False
2        False
3        False
4        False
         ...  
16951    False
16952    False
16953    False
16954    False
16955    False
Length: 16681, dtype: bool

In [25]:
# ВЫявление дубликатов по соответсующим столбцам
df.duplicated(subset = ['name', 'platform', 'year_of_release'], keep='first')

0        False
1        False
2        False
3        False
4        False
         ...  
16951    False
16952    False
16953    False
16954    False
16955    False
Length: 16681, dtype: bool

Вывод: 
Не неадено дублирующих строк, но найдены дублирующие значения в столбцах genre и rating. Требуется привести данные в столбцах к соответствующим регистрам: с заглавной буквы и полностью заглавные буквы соответственно.

In [26]:
for column in ['platform','year_of_release','genre','rating']:
    print(f'Уникальные значения в столбце {column}:')
    print(df[column].sort_values().unique())
    print()

Уникальные значения в столбце platform:
['2600' '3DO' '3DS' 'DC' 'DS' 'GB' 'GBA' 'GC' 'GEN' 'GG' 'N64' 'NES' 'NG'
 'PC' 'PCFX' 'PS' 'PS2' 'PS3' 'PS4' 'PSP' 'PSV' 'SAT' 'SCD' 'SNES' 'TG16'
 'WS' 'Wii' 'WiiU' 'X360' 'XB' 'XOne']

Уникальные значения в столбце year_of_release:
[1980. 1981. 1982. 1983. 1984. 1985. 1986. 1987. 1988. 1989. 1990. 1991.
 1992. 1993. 1994. 1995. 1996. 1997. 1998. 1999. 2000. 2001. 2002. 2003.
 2004. 2005. 2006. 2007. 2008. 2009. 2010. 2011. 2012. 2013. 2014. 2015.
 2016.]

Уникальные значения в столбце genre:
['ACTION' 'ADVENTURE' 'Action' 'Adventure' 'FIGHTING' 'Fighting' 'MISC'
 'Misc' 'NS' 'PLATFORM' 'PUZZLE' 'Platform' 'Puzzle' 'RACING'
 'ROLE-PLAYING' 'Racing' 'Role-Playing' 'SHOOTER' 'SIMULATION' 'SPORTS'
 'STRATEGY' 'Shooter' 'Simulation' 'Sports' 'Strategy']

Уникальные значения в столбце rating:
['AO' 'E' 'E10+' 'EC' 'K-A' 'M' 'NS' 'RP' 'T']



In [27]:
df = df[~df.duplicated()]
df

name platform  year_of_release         genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       na_sales eu_sales jp_sales  other_sales  critic_score  user_score  \
0         41.36    28.96     3.77         8.45          76.0         8.0   
1         29.08     3.58     6.81         0.77          -1.0        -1.0   
2         15.68    12.76     3.79         3.29          82.0         8.3   
3         15.61    10.93     3.28         2.95          80.0         8.0   
4         11.27     8.89    10.22         1.00          -1.0        -1.0   
...         ...      ...      ...          ...           ...         ...   
16951      0.00      0.0     0.01         0.00          -1.0        -1.0   
16952      0.00     0.01      0.0         0.00          -1.0        -1.0   
16953      0.00      0.0     0.01         0.00          -1.0        -1.0   
16954      0.01      0.0      0.0         0.00          -1.0        -1.0   
16955      0.00      0.0     0.01         0.00          -1.0        -1.0   

      rating  
0          E  
1         NS  
2          E  
3          E  
4         NS  
...      ...  
16951     NS  
16952     NS  
16953     NS  
16954     NS  
16955     NS  

[16502 rows x 11 columns]

In [28]:
# Избавляемся от дубликатов в столбцах
df.loc[:, 'genre'] = df['genre'].str.title()
df.loc[:, 'rating'] = df['rating'].str.upper()
df

name platform  year_of_release         genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       na_sales eu_sales jp_sales  other_sales  critic_score  user_score  \
0         41.36    28.96     3.77         8.45          76.0         8.0   
1         29.08     3.58     6.81         0.77          -1.0        -1.0   
2         15.68    12.76     3.79         3.29          82.0         8.3   
3         15.61    10.93     3.28         2.95          80.0         8.0   
4         11.27     8.89    10.22         1.00          -1.0        -1.0   
...         ...      ...      ...          ...           ...         ...   
16951      0.00      0.0     0.01         0.00          -1.0        -1.0   
16952      0.00     0.01      0.0         0.00          -1.0        -1.0   
16953      0.00      0.0     0.01         0.00          -1.0        -1.0   
16954      0.01      0.0      0.0         0.00          -1.0        -1.0   
16955      0.00      0.0     0.01         0.00          -1.0        -1.0   

      rating  
0          E  
1         NS  
2          E  
3          E  
4         NS  
...      ...  
16951     NS  
16952     NS  
16953     NS  
16954     NS  
16955     NS  

[16502 rows x 11 columns]

In [29]:
# Проверка
df['genre'].nunique() # должно быть 13

13

In [30]:
# Проверка
df['rating'].nunique() # должно быть 9

9

В процессе подготовки данных вы могли удалять данные, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях. После проведения предобработки данных напишите общий промежуточный вывод.

In [31]:
# Абсолютное количество удаленных дубликатов
absolut_duble = rows_cnt - len(df)
absolut_duble

454

In [32]:
# Абсолютное количество удаленных дубликатов
print(f"{round(absolut_duble / len(df) * 100, 2)}%")

2.75%


ВЫВОД
В результате предобработки данных была проведена работа над пропусками и дубликатами. Дубликатов в данных не было обнаружено, количество пропусков 454, что составляет менее 3% от весх данных в таблице.

## Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например df_actual.

In [33]:
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)]
df_actual

name platform  \
0                                            Wii Sports      Wii   
2                                        Mario Kart Wii      Wii   
3                                     Wii Sports Resort      Wii   
6                                 New Super Mario Bros.       DS   
7                                              Wii Play      Wii   
...                                                 ...      ...   
16947                     Men in Black II: Alien Escape       GC   
16949                Woody Woodpecker in Crazy Castle 5      GBA   
16950  SCORE International Baja 1000: The Official Game      PS2   
16952                                  LMA Manager 2007     X360   
16954                                  Spirits & Spells      GBA   

       year_of_release     genre  na_sales eu_sales jp_sales  other_sales  \
0               2006.0    Sports     41.36    28.96     3.77         8.45   
2               2008.0    Racing     15.68    12.76     3.79         3.29   
3               2009.0    Sports     15.61    10.93     3.28         2.95   
6               2006.0  Platform     11.28     9.14      6.5         2.88   
7               2006.0      Misc     13.96     9.18     2.93         2.84   
...                ...       ...       ...      ...      ...          ...   
16947           2003.0   Shooter      0.01      0.0      0.0         0.00   
16949           2002.0  Platform      0.01      0.0      0.0         0.00   
16950           2008.0    Racing      0.00      0.0      0.0         0.00   
16952           2006.0    Sports      0.00     0.01      0.0         0.00   
16954           2003.0  Platform      0.01      0.0      0.0         0.00   

       critic_score  user_score rating  
0              76.0         8.0      E  
2              82.0         8.3      E  
3              80.0         8.0      E  
6              89.0         8.5      E  
7              58.0         6.6      E  
...             ...         ...    ...  
16947          -1.0        -1.0      T  
16949          -1.0        -1.0     NS  
16950          -1.0        -1.0     NS  
16952          -1.0        -1.0     NS  
16954          -1.0        -1.0     NS  

[12829 rows x 11 columns]

## Категоризация данных

Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [34]:
df_actual = df_actual.copy()
def cus_rating_func (row):
    if (row['user_score'] >= 8) & (row['user_score'] <= 10):
        return 'высокая оценка'
    elif (row['user_score'] >= 3) & (row['user_score'] < 8):
        return 'средняя оценка'
    elif (row['user_score'] >= 0) & (row['user_score'] < 3):
        return 'низкая оценка'
    else: 'NS'
df_actual.loc[:, 'user_category'] = df_actual.apply(cus_rating_func, axis = 1)
df_actual

name platform  \
0                                            Wii Sports      Wii   
2                                        Mario Kart Wii      Wii   
3                                     Wii Sports Resort      Wii   
6                                 New Super Mario Bros.       DS   
7                                              Wii Play      Wii   
...                                                 ...      ...   
16947                     Men in Black II: Alien Escape       GC   
16949                Woody Woodpecker in Crazy Castle 5      GBA   
16950  SCORE International Baja 1000: The Official Game      PS2   
16952                                  LMA Manager 2007     X360   
16954                                  Spirits & Spells      GBA   

       year_of_release     genre  na_sales eu_sales jp_sales  other_sales  \
0               2006.0    Sports     41.36    28.96     3.77         8.45   
2               2008.0    Racing     15.68    12.76     3.79         3.29   
3               2009.0    Sports     15.61    10.93     3.28         2.95   
6               2006.0  Platform     11.28     9.14      6.5         2.88   
7               2006.0      Misc     13.96     9.18     2.93         2.84   
...                ...       ...       ...      ...      ...          ...   
16947           2003.0   Shooter      0.01      0.0      0.0         0.00   
16949           2002.0  Platform      0.01      0.0      0.0         0.00   
16950           2008.0    Racing      0.00      0.0      0.0         0.00   
16952           2006.0    Sports      0.00     0.01      0.0         0.00   
16954           2003.0  Platform      0.01      0.0      0.0         0.00   

       critic_score  user_score rating   user_category  
0              76.0         8.0      E  высокая оценка  
2              82.0         8.3      E  высокая оценка  
3              80.0         8.0      E  высокая оценка  
6              89.0         8.5      E  высокая оценка  
7              58.0         6.6      E  средняя оценка  
...             ...         ...    ...             ...  
16947          -1.0        -1.0      T            None  
16949          -1.0        -1.0     NS            None  
16950          -1.0        -1.0     NS            None  
16952          -1.0        -1.0     NS            None  
16954          -1.0        -1.0     NS            None  

[12829 rows x 12 columns]

Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [35]:
df_actual = df_actual.copy()
def cr_rating_func (rows):
    if (rows['critic_score'] >= 80) & (rows['critic_score'] <= 100):
        return 'высокая оценка'
    elif (rows['critic_score'] >= 30) & (rows['critic_score'] < 80):
        return 'средняя оценка'
    elif (rows['critic_score'] >= 0) & (rows['critic_score'] < 30):
        return 'низкая оценка'
    else: 'NS'
df_actual.loc[:, 'critic_category'] = df_actual.apply(cr_rating_func, axis = 1)
df_actual

name platform  \
0                                            Wii Sports      Wii   
2                                        Mario Kart Wii      Wii   
3                                     Wii Sports Resort      Wii   
6                                 New Super Mario Bros.       DS   
7                                              Wii Play      Wii   
...                                                 ...      ...   
16947                     Men in Black II: Alien Escape       GC   
16949                Woody Woodpecker in Crazy Castle 5      GBA   
16950  SCORE International Baja 1000: The Official Game      PS2   
16952                                  LMA Manager 2007     X360   
16954                                  Spirits & Spells      GBA   

       year_of_release     genre  na_sales eu_sales jp_sales  other_sales  \
0               2006.0    Sports     41.36    28.96     3.77         8.45   
2               2008.0    Racing     15.68    12.76     3.79         3.29   
3               2009.0    Sports     15.61    10.93     3.28         2.95   
6               2006.0  Platform     11.28     9.14      6.5         2.88   
7               2006.0      Misc     13.96     9.18     2.93         2.84   
...                ...       ...       ...      ...      ...          ...   
16947           2003.0   Shooter      0.01      0.0      0.0         0.00   
16949           2002.0  Platform      0.01      0.0      0.0         0.00   
16950           2008.0    Racing      0.00      0.0      0.0         0.00   
16952           2006.0    Sports      0.00     0.01      0.0         0.00   
16954           2003.0  Platform      0.01      0.0      0.0         0.00   

       critic_score  user_score rating   user_category critic_category  
0              76.0         8.0      E  высокая оценка  средняя оценка  
2              82.0         8.3      E  высокая оценка  высокая оценка  
3              80.0         8.0      E  высокая оценка  высокая оценка  
6              89.0         8.5      E  высокая оценка  высокая оценка  
7              58.0         6.6      E  средняя оценка  средняя оценка  
...             ...         ...    ...             ...             ...  
16947          -1.0        -1.0      T            None            None  
16949          -1.0        -1.0     NS            None            None  
16950          -1.0        -1.0     NS            None            None  
16952          -1.0        -1.0     NS            None            None  
16954          -1.0        -1.0     NS            None            None  

[12829 rows x 13 columns]

Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [36]:
top_platforms = df_actual['platform'].value_counts().head(7).reset_index()
top_platforms

platform  count
0      PS2   2131
1       DS   2126
2      Wii   1279
3      PSP   1184
4     X360   1126
5      PS3   1090
6      GBA    813

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

### Вывод

Были проанализированы данные о видеоиграх и выделены топ-7 платформ по количеству выпущенных игр

Описание проделанной работы
- Создание датасета: Сформировали DataFrame с названиями игр и их платформами и пороанализировали пропуски и дубликаты.
- Добавление категории оценок: Ввели новый столбец user_category, классифицирующий оценки на три категории (высокая, средняя, низкая).
- Анализ платформ: Подсчитали количество игр на каждой платформе и выделили топ-7.
- Отображение результатов: Использовали tabulate или PrettyTable для красивого вывода таблицы.
Описание среза данных и новых полей
Срез данных: Анализировались платформы с наибольшим количеством игр.
Новые поля:
user_category: Классификация оценок.
Топ-7 платформ: DataFrame с платформами и количеством игр.
Таким образом, мы выделили популярные платформы и добавили полезные категории оценок для дальнейшего анализа.